independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java

```diff
@@ -55,8 +73,30 @@ public void walk(PathVisitor visitor) {
         PathTreeVisit.walk(dir, dir, pathFilter, getMultiReleaseMapping(), visitor);
     }
 
+    private String ensureRelative(String path) {
+        if (path == null) {
+            return path;
+        }
+        if (isAbsolutePath(path)) {
+            final Path absolutePath = dir.getFileSystem().getPath(path).normalize().toAbsolutePath();
+            if (absolutePath.startsWith(dir)) {
+                return dir.relativize(absolutePath).toString();
+            }
+            return null;
+        }
+        if (path.contains("..")) {
+            final Path absolutePath = dir.resolve(path).normalize().toAbsolutePath();
+            if (absolutePath.startsWith(dir)) {
+                return dir.relativize(absolutePath).toString();
+            }
+            return null;
+        }
```

Comment:
I really don't think we should do this... In my opinion, the string being passed as input should be interpreted as a paths to a resource in the classpath, i.e. something that we'd normally pass to `Class#getResource`.

The goal of `DirectoryPathTree` as I understand it is to convert such "path in the classpath" to an actual path on the filesystem.

With that in mind, it doesn't make sense to allow input strings to include the path to `src/main/resources` on the filesystem (like you did further up, see my other comment), or to interpret things such as `..` which are a filesystem concept (and not something that is usually supported for paths to resources in the classpath).

Reply:
About this one, I think I'd keep it. The thing is that the PathTree is also exposed as an actual file system path tree API in other places. However, we still should not allow reading outside the tree root.

Reply:
I do agree it's ugly though.

Reply:
Supporting `..` does raise a problem, though... But maybe one that was pre-existing.

If I pass `foo/bar/..` to `DirectoryPathTree`, it will automatically be resolved to `foo/`. So in a build step for example, we might end up accepting the path as valid.
Later at runtime, we might pass the same string to `Class#getRessource`. And then it will fail, because `..` has no special meaning for `Class#getResource`...

Maybe a better behavior would be to simple escape `..`, so that it's interpreted as a literal folder named `..`? And if that's not possible, maybe we should just reject any path with a component named `..`?

Damn, this is starting to look dodgy. I sure hope there aren't many similar pitfalls with Windows paths.

Reply:
Yes, there could be path element name `a..b`

Reply:
The issue is that, this API isn't strictly limited to classpath resources, which is why I am hesitating.

Reply:
I think especially if you are using `ApplicationArchive.getChildPath(path)`, I don't think it's agreed upon that the `path` argument is actually following classloader resource naming format. I think it used be an actual relative FS path before `PathTree` was introduced.
The `PathTree` is also used to represent artifact content of application dependencies.

Reply:
Hu. I didn't know that.

So we have a single API for multiple types of resource trees, some of which don't have the same syntax for their path strings? That looks like a problem in itself...

Ok, I'll trust you to pick what seems best to you, but to me those are the only possible solutions in the scope of this PR:

* Your implementation, which allows using `..`, which could lead to incorrectly accepting invalid paths for classpath resource trees (see my example above).
* My implementation, which prevents using `..`, which could lead to incorrectly rejecting valid path for non-classpath resource trees.

Alternatively, we could provide a flag in the constructor, to switch between the two behaviors.

Reply:
Yes, before `PathTree` we would simple do `path.resolve(childPath)` in `ApplicationArchive.getChildPath(childPath)`. The primary goal for `PathTree` wasn't actually to represent the classpath resources but to provide a single API to process directory and archive content with the option of applying filters. I would probably keep it that way.
However, I get your point. We actually do have `PathTreeClassPathElement` which implements `ClassPathElement` and its purpose is pretty much what you are asking for. But it's not yet exposed from `ApplicationArchive`. I think it should be, it would have to be a separate refactoring task.
You can actually use `QuarkusClassLoader.getElements(resourceName, localOnly)` which will return a list of `PathTreeClassPathElement`s. But it was added kind of quickly and isn't a nice/proper API for classpath resource locating. Also, it's different from the `ApplicationArchive`, in a sense that `ApplicationArchive`s are a subset of all the dependencies that are identified as such.

Reply:
Alright, I approved the PR; I'm not sure what you're going to do with this `..` handling, so I'll let you merge once you're done?

Thanks again for all the work.

Reply:
I am thinking I probably wouldn't mind the change you proposed. It wouldn't be a huge limitation at this point and we could reconsider supporting `..` in case we really need to one day. Would you like to add this change to your original PR? Thank you @yrodiere

Reply:
Will do. Thanks.